operator/watchers/service_sync.go

```diff
@@ -0,0 +1,294 @@
+	ForDeletion(svc *slim_corev1.Service) (out *serviceStore.ClusterService)
+}
+
+type ServiceSyncParams struct {
+	cell.In
+
+	Config                  ServiceSyncConfig
+	Log                     *slog.Logger
+	ClusterInfo             cmtypes.ClusterInfo
+	Clientset               k8sClient.Clientset
+	Services                resource.Resource[*slim_corev1.Service]
+	Endpoints               resource.Resource[*k8s.Endpoints]
+	StoreFactory            store.Factory
+	ClusterServiceConverter ClusterServiceConverter
+}
+
+type serviceSync struct {
+	ServiceSyncParams
+
+	storePromise promise.Promise[store.SyncStore]
+}
+
+func registerServiceSync(jg job.Group, p ServiceSyncParams) {
+	if !p.Config.Enabled || !p.Clientset.IsEnabled() {
+		return
+	}
+
+	s := &serviceSync{ServiceSyncParams: p}
+
+	storeResolver, storePromise := promise.New[store.SyncStore]()
+	s.storePromise = storePromise
+
+	jg.Add(
+		job.OneShot(
+			"service-sync",
+			s.loop,
+		),
+		job.OneShot(
+			"run-store",
+			func(ctx context.Context, health cell.Health) error {
+				var backend kvstore.BackendOperations
+				if s.Config.Backend != nil {
+					var err error
+					backend, err = s.Config.Backend.Await(ctx)
+					if err != nil {
+						storeResolver.Reject(err)
+						return err
+					}
+				} else {
+					backend = kvstore.Client()
+				}
+				store := s.StoreFactory.NewSyncStore(
+					s.ClusterInfo.Name,
+					backend,
+					serviceStore.ServiceStorePrefix,
+				)
+				storeResolver.Resolve(store)
+				store.Run(ctx)
+				return nil
+			},
+		),
+	)
+}
+
+func (s *serviceSync) loop(ctx context.Context, health cell.Health) error {
+	store, err := s.storePromise.Await(ctx)
+	if err != nil {
+		return err
+	}
+
+	converter := s.ClusterServiceConverter
+
+	services, err := s.Services.Store(ctx)
+	if err != nil {
+		return err
+	}
+
+	endpoints, err := s.Endpoints.Store(ctx)
+	if err != nil {
+		return err
+	}
+
+	getEndpoints := func(namespace, name string) []*k8s.Endpoints {
+		eps, _ := endpoints.ByIndex(operatorK8s.ServiceIndex, namespace+"/"+name)
+		return eps
+	}
+
+	serviceEvents := s.Services.Events(ctx)
+	endpointEvents := s.Endpoints.Events(ctx)
+
+	upsert := func(cs *serviceStore.ClusterService) {
+		if err := store.UpsertKey(ctx, cs); err != nil {
+			// An error is triggered only in case it concerns service marshaling,
+			// as kvstore operations are automatically re-tried in case of error.
+			s.Log.Warn("Failed synchronizing service",
+				logfields.Error, err,
+				logfields.K8sSvcName, cs.Name,
+				logfields.K8sNamespace, cs.Namespace,
+			)
+		}
+	}
+
+	for serviceEvents != nil || endpointEvents != nil {
+		select {
+		case ev, ok := <-serviceEvents:
+			if !ok {
+				serviceEvents = nil
+				continue
+			}
+			ev.Done(nil)
+
+			// Skip headless services.
+			if ev.Kind != resource.Sync && isHeadless(ev.Object) {
+				continue
+			}
```

Comment:
I don't think that headless services should be skipped. It is just that they should have no frontend, rather than one with the `None` address.

Reply:
They were most likely dropped here in the previous service cache implementation:

https://github.com/cilium/cilium/blob/2d5639f5d8e429ebe8b6e917356c507036ec5da1/pkg/k8s/service.go#L199-L214

Reply:
Ah you're right. `watchers/service.go` skips them but shouldn't do that here.